src/main/kotlin/coverosR3z/server/Server.kt

```diff
@@ -84,56 +76,78 @@ class Server(val port: Int, private val dbDirectory: String? = null) {
+            val possibleOptions = listOf(
+                OptionGroup("-d", directoryOption),
+                OptionGroup("--dbdirectory", directoryOption),
+                OptionGroup("-p", portOption),
+                OptionGroup("--port", portOption),
+                OptionGroup("--no-disk-persistence", diskPersistenceOption))
+
             val fullInput = args.joinToString(" ")
             return if (args.isEmpty() || args[0].isBlank()) {
                 ServerOptions()
             } else {
-                //first boolean in each Pair indicates whether the flag is specified
-                var port = Pair<Boolean, Int?>(false, null)
-                var db = Pair<Boolean, String?>(false, null)
-                var ndp = Pair<Boolean, Boolean?>(false, null)
-
-                args.forEachIndexed { i, it ->
-
-                    if (it == "--no-disk-persistence") {
-                        if (ndp.first) throw ServerOptionsException("The disk persistence option was specified multiple times. This is not allowed, go to jail.")
-                        ndp = Pair(true, true)
-                    } else if (it.startsWith("-p")) {
-                        if (port.first) throw ServerOptionsException("Multiple port values were provided. This is not allowed, go to jail.")
-                        val portStr: String =
-                            when {
-                                it.length > 2 -> it.substring(2)
-                                i + 1 < args.size -> args[i + 1]
-                                else -> throw ServerOptionsException("The port option was specified, but no port number was given. This is dumb, you are dumb.")
-                            }
-                        validatePort(portStr, fullInput)
-                        port = Pair(true, portStr.toInt())
-                    } else if (it.startsWith("-d")) {
-                        if (db.first) throw ServerOptionsException("The database option was specified multiple times. This is not allowed, go to jail.")
-                        val dbStr: String =
-                            when {
-                                it.length > 2 -> it.substring(2)
-                                i + 1 < args.size && validateDir(args[i + 1]) -> args[i + 1]
-                                else -> throw ServerOptionsException("The directory option was provided without a directory value")
-                            }
-                        db = Pair(true, dbStr)
-                    } else if (it.startsWith("-h")) {
-                        throw ServerOptionsException(fullHelpMessage)
-                    } else if (it.startsWith("-")) {
-                        if (it.length >= 2) {
-                            if (it[1] !in setOf('h', 'd', 'p')) {
-                                throw ServerOptionsException("Unrecognized option(s)")
-                            }
+                var currentIndex = 0
+
+                loop@ while (currentIndex < args.size) {
```

Comment:
goto considered harmless

Reply:
I am near certain there is a for index in args.indices but whatever floats yer boat. Well actually, that might be advisable best practice...

Reply:
But I'm not stepping through incrementally, so cannot use a for loop

Reply:
yeah, so it would seem. I'd say this is verging on pretty hard to track, but the options classes are big improvements